Solved the task

Likh-Alex opened the pull request:
- MySQL connection is established through `ConnectionUtil` class
- Created `ManufacturerDaoJdbcImpl` with all required `CRUD` methods
- created custom exception class `DataProcessingException`
- created `init_db.sql` file with scripts to create `taxi_service` schema and `manufacturers` table

Likh-Alex committed:
Solved the task

Ketch92 commented on pom.xml:
```diff
@@ -18,6 +18,13 @@
         </maven.checkstyle.plugin.configLocation>
     </properties>
 
```
add an empty line below or remove this empty line to separate dependencies

Likh-Alex replied:
I have added empty lines to separate the dependencies.

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+                + " VALUES (?, ?)";
+
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection
+                    .prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
```
move preparedstatement to try with resorces together with connection

Likh-Alex replied:
I got it - so instead of implicitly calling `statement.close()` we can put also inside the `try` block and it will force `preparedStatement` to invoke `close()`, same as with exceptions)

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection
+                    .prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            String name = manufacturer.getName();
+            String country = manufacturer.getCountry();
```
no need to declare new variables, just pass manufacturer.getName() to statement.setString

Likh-Alex replied:
Done

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+            preparedStatement.executeUpdate();
+            ResultSet resultSet = preparedStatement.getGeneratedKeys();
+            if (resultSet.next()) {
+                manufacturer.setId(resultSet.getObject(1, Long.class));
```
it's bad to access the column by index and not by name. Just in case if someone later adds a new column to the table, then ordering changes and the code then crashes due to wrong column numbering

Likh-Alex replied:
You are right, refactored to access the column by `manufacturer_id`

Likh-Alex replied:
But - when I try to run the code with accessing the column by `columnLabel` - I got a custom exception. Even though new `manufacturer` get created in the db. I will keep calling by `index` for now.

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+            if (resultSet.next()) {
+                manufacturer.setId(resultSet.getObject(1, Long.class));
+            }
+            preparedStatement.close();
+            resultSet.close();
```
if you add prepareStatement to try with resources then this is redundant

Likh-Alex replied:
removed redundant `closes()` at the moment when moved `preparedStatements` inside the `try` block

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+            preparedStatement.close();
+            resultSet.close();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can not add new manufacturer"
```
space in the end of string literal would be good

Likh-Alex replied:
added a space

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+                + "AND deleted = FALSE;";
+
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
```
same, move to try with resources

Likh-Alex replied:
done

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+        String query = "SELECT * FROM manufacturers WHERE deleted = FALSE;";
+        List<Manufacturer> manufacturerList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
```
same

Likh-Alex replied:
Done here as well

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+                + " AND deleted = FALSE;";
+
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection
+                    .prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
```
same

Likh-Alex replied:
and here also done

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+            preparedStatement.setString(2, manufacturer.getCountry());
+            preparedStatement.setLong(3, manufacturer.getId());
+            preparedStatement.executeUpdate();
+            ResultSet resultSet = preparedStatement.getGeneratedKeys();
+            while (resultSet.next()) {
+                manufacturer = createManufacturer(resultSet);
+                preparedStatement.close();
+                resultSet.close();
+            }
```
I'm not sure if it's needed as you pass fully formed manufacturer

Likh-Alex replied:
You are right - if we would not update the `manufacturer` we will catch an `exception`, otherwise the update was successful and we can return the input `manufacturer`.

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+        return manufacturer;
+    }
+
+    private Manufacturer createManufacturer(ResultSet resultSet) throws SQLException {
+        try {
+            long manufacturerId = resultSet.getObject("manufacturer_id", Long.class);
+            String manufacturerCountry = resultSet.getObject("manufacturer_country", String.class);
+            String manufacturerName = resultSet.getObject("manufacturer_name", String.class);
+            Manufacturer manufacturer = new Manufacturer(manufacturerName, manufacturerCountry);
+            manufacturer.setId(manufacturerId);
+            return manufacturer;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can not parse data"
+                    + " from manufacturer with id: "
+                    + resultSet.getObject("manufacturer_id", Long.class), e);
+        }
+    }
```
put this private method down below all public methods

Likh-Alex replied:
I think it is better to keep near the methods were I use it. Also moved it closer to the related methods

Ketch92 commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,144 @@
+                + " SET deleted = TRUE"
+                + " WHERE manufacturer_id = ? ";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection
+                    .prepareStatement(query);
```
same

Likh-Alex replied:
Also move into the `try` block

Likh-Alex committed:
Refactoring
- Moved all `preparedStatement` into try with resources
- removed redundant `createManufacturer` in `update` method

RomanDubovskyi commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,133 @@
+            if (resultSet.next()) {
+                manufacturer.setId(resultSet.getObject(1, Long.class));
+            }
+            resultSet.close();
```
it's a bit redundant to close result set. If statement is closed correctly, resultSet will be closed as well
https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/Statement.html#close()

Likh-Alex replied:
Removed, reviewed the article - now it's clear to me. Thank you

RomanDubovskyi commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,133 @@
+            if (resultSet.next()) {
+                Optional<Manufacturer> newManufacturer = Optional
+                        .of(createManufacturer(resultSet));
+                resultSet.close();
```
same here, we can make the code a bit shorter

Likh-Alex replied:
Done as well

RomanDubovskyi commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,133 @@
+        return manufacturerList;
+    }
+
+    private Manufacturer createManufacturer(ResultSet resultSet) throws SQLException {
```
```suggestion
    private Manufacturer createManufacturer(ResultSet resultSet) {
```
if we handle exception inside of the method with a try catch construction we don't need to use throws as well

Likh-Alex replied:
Removed `throws` from the method signature.

RomanDubovskyi commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,133 @@
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection
+                        .prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {
```
```suggestion
                        .prepareStatement(query)) {
```
we don't need to return generated keys here

Likh-Alex replied:
Removed.

Likh-Alex committed:
Refactoring
- removed `resultSet.close()` as redundant expression
- refactored `createManufacturer` method - removed `throws DataProcessingException` from the method signature

RomanDubovskyi commented on src/main/resources/init_db.sql:
```diff
@@ -0,0 +1,12 @@
+    `deleted`              tinyint(1)   NOT NULL DEFAULT '0',
+    PRIMARY KEY (`manufacturer_id`)
+) ENGINE = InnoDB
+  AUTO_INCREMENT = 7
```
why 7 ?

Likh-Alex replied:
I wish to know, it is a copy of `CREATE` script from MySQL workbench.

RomanDubovskyi commented on src/main/java/taxi/dao/jdbcimpl/ManufacturerDaoJdbcImpl.java:
```diff
@@ -0,0 +1,127 @@
+
+    private Manufacturer createManufacturer(ResultSet resultSet) {
+        try {
+            long manufacturerId = resultSet.getObject("manufacturer_id", Long.class);
```
```suggestion
            Long manufacturerId = resultSet.getObject("manufacturer_id", Long.class);
```

Likh-Alex replied:
Surely will do

RomanDubovskyi reviewed (approved):
nice) 
don't forget to change primitive to wrapper before merge

Likh-Alex committed:
raplaced type of `manufacturer_id` witj `Long`
